module/zfs/zvol.c

```diff
@@ -1008,6 +1008,16 @@ zvol_request(struct request_queue *q, struct bio *bio)
 				zvol_write(zvr);
 		}
 	} else {
+		/*
+		 * rangelock_add_reader() can't handle overlap of two ranges
+		 * with the same offset when either has length 0, so return
+		 * without entering rangelock if it's a read bio with size 0.
+		 */
```

Comment:
This fix looks good to me, but how about rephrasing the comment a little to explain how a size of zero can happen.  I think that would better get at the heart of the issue since the zero length is the unusual condition.  The range lock code itself handled this condition as expected.  How about something like:

```c
/*
 * The SCST driver, and possibly others, may issue READ IOs with a
 * length of zero bytes.  These empty IOs contain no data and require
 * no additional handling.
 */
```

Reply:
ok, will change to it.

Reply:
Updated via force push.